### PR TITLE
Refs and decompositions for index_{add,copy,select,fill}

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #85002

As per title

### CODE REVIEW

**Comment by jansel** [comment]
How does the performance of this compare to eager using TorchInductor?

**Comment by Lezcano** [comment]
The generated code for the lowering of `index_select` and its decomposition is the same. I think we should remove the lowering.

**Comment by Lezcano** [comment]
The performance is the same:
```python
[------------------- index_select -------------------]
                  |  Decomposed  |  Lowering  |  Eager
32 threads: ------------------------------------------
      (256, 256)  |      20      |     18     |    18 
      (256, 512)  |      18      |     18     |    20 
      (512, 256)  |      18      |     18     |    18 
      (512, 512)  |      18      |     18     |    18 

Times are in microseconds (us).
```
Note that the first two columns have the same triton implementation, so that swing between 20 and 18 is just variance.

**Comment by Lezcano** [comment]
I can benchmark `index_add` and `index_copy` against eager later on.

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -1547,6 +1547,48 @@ def maybe_mask(vals, length, range_max, adaptive, dim):
...
+    return x
+
+
+@register_decomposition(aten.index_add_)
+def index_add_(x, dim, index, tensor, *, alpha=1):
+    dim = utils.canonicalize_dims(x.ndim, dim)
+    utils.check(index.ndim <= 1, lambda: f"Index should have dimension 1 or 0 (got {index.ndim})")
+    utils.check(utils.is_integer_dtype(index.dtype), lambda: "Expected dtype int32 or int64 for index")
+    idx = (slice(None),) * dim + (index,)
+    x[idx] += alpha * tensor
```
Comment (jansel) [inline torch/_decomp/decompositions.py:1606]: Isn't `x[idx] +=` the same as `index_add_`?

Comment (Lezcano) [reply torch/_decomp/decompositions.py:1606]: The general `x[idx] +=` is implemented via `index_put_(..., accumulate=True)`. Same for `x[idx] =` but with `accumulate=False`.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 2]
```diff
@@ -1095,8 +1095,6 @@ Tensor & index_select_out_cpu_(const Tensor & self, int64_t dim, const Tensor &
   TORCH_CHECK(index.scalar_type() == ScalarType::Long || index.scalar_type() == ScalarType::Int, "index_select(): Expected dtype int32 or int64 for index");
   TORCH_CHECK(self.scalar_type() == result.scalar_type(),
               "index_select(): self and result must have the same scalar type");
-  TORCH_CHECK(dim == 0 || dim < self.dim(),
-              "index_select(): Indexing dim ", dim, " is out of bounds of tensor");
```
Comment (Lezcano) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:1099]: Not needed as we test it in `wrap_dim`.

**Comment by Lezcano** [comment]
In `index_add` we're also doing reasonably better than eager:
```
[---------------- index_add -----------------]
                       |  Decomposed  |  Eager
32 threads: ----------------------------------
      (256, 256, 256)  |     1.15     |    1.7
      (256, 256, 512)  |     2.30     |    3.5
      (512, 512, 256)  |     4.55     |    7.3
      (512, 512, 512)  |     9.12     |   15.1

Times are in milliseconds (ms).
```

Now, we end up generating two kernels, which is not great.
```python
    def forward(self, x_1, idx_1, src_1):
                                                     
        # No stacktrace found for following nodes 
        index = torch.ops.aten.index.Tensor(x_1, [idx_1])
        mul = torch.ops.aten.mul.Tensor(src_1, 1);  src_1 = None
        add_ = torch.ops.aten.add_.Tensor(index, mul);  index = mul = None
        index_put_ = torch.ops.aten.index_put_.default(x_1, [idx_1], add_);  x_1 = idx_1 = add_ = None
        return (index_put_,)
```
and cleaning a bit the code
```python
def forward(x, idx, src):
    index = torch.ops.aten.index.Tensor(x, [idx])
    return torch.ops.aten.index_put_.default(x, [idx], index + src);
```
This is not great, as it needs to do two operations, rather than using `index_put_(accumulate=True)`. Do you know what's going on here @ngimel? Is this an oversight in core?

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -1547,6 +1547,66 @@ def maybe_mask(vals, length, range_max, adaptive, dim):
...
+def index_copy_(x, dim, index, tensor):
+    dim = utils.canonicalize_dims(x.ndim, dim)
+    utils.check(
+        index.ndim <= 1,
+        lambda: f"Index should have dimension 1 or 0 (got {index.ndim})",
+    )
+    utils.check(
+        utils.is_integer_dtype(index.dtype),
+        lambda: "Expected dtype int32 or int64 for index",
+    )
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/_decomp/decompositions.py:1588]: coerces. In all fairness, this check could probably go, as it should be done in the implementation of advanced indexing already.

**Comment by Lezcano** [comment]
I think what's happening here is that `x[idx] += tensor` in Python it's decomposed in `y = x[idx]; y += tensor, x[idx] = y`, as 
```python
def f(a, b, c):
    a[:, b] += c
```
compiles into
```python
  7           0 LOAD_FAST                0 (a)
              2 LOAD_CONST               0 (None)
              4 LOAD_CONST               0 (None)
              6 BUILD_SLICE              2
              8 LOAD_FAST                1 (b)
             10 BUILD_TUPLE              2
             12 DUP_TOP_TWO
             14 BINARY_SUBSCR
             16 LOAD_FAST                2 (c)
             18 INPLACE_ADD
             20 ROT_THREE
             22 STORE_SUBSCR
             24 LOAD_CONST               0 (None)
             26 RETURN_VALUE
```
As such, the decomposition into `x[idx]` is not great. By using `aten.index_put` directly together with `accumulate=True`, we manage to almost make the kernel go 2x as fast, and we get to be up to 3x times faster than eager!
```
[---------------- index_add -----------------]
                       |  Decomposed  |  Eager
32 threads: ----------------------------------
      (256, 256, 256)  |      722     |   1691
      (256, 256, 512)  |     1362     |   3521
      (512, 512, 256)  |     2726     |   7290
      (512, 512, 512)  |     5449     |  15100

Times are in microseconds (us).
```
Compare these with those from https://github.com/pytorch/pytorch/pull/85002#issuecomment-1247188844.

File: torch/_decomp/decompositions.py (hunk 1 above)
Comment (Lezcano) [reply torch/_decomp/decompositions.py]: Nb. The comment above is not true. See https://github.com/pytorch/pytorch/pull/85002#issuecomment-1247188844 and https://github.com/pytorch/pytorch/pull/85002#issuecomment-1248524492

**Comment by ngimel** [comment]
Awesome, so augmented assignment magic happens at the python level, and yeah, calling index_put directly is the way to go. Too bad it doesn't extend to -=, *= etc ;-)

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -2944,6 +2951,108 @@ def unbind(t: TensorLikeType, dim: int = 0) -> TensorSequenceType:
...
+@register_decomposition(torch.ops.aten.index_copy)
+@out_wrapper()
+def index_copy(x: TensorLike, dim: int, index: TensorLike, tensor: TensorLike):
+    return x.clone().index_copy_(dim, index, tensor)
+
+
+# The decomposition of this function dispatches to aten.index_put_ for efficiency
+# We cannot do that in Python, as torch.index_put_ does not support slice(None)s See
+# https://github.com/pytorch/pytorch/pull/85002#issuecomment-1248524492
+def index_copy_(x: TensorLike, dim: int, index: TensorLike, tensor: TensorLike):
```
Comment (ngimel) [inline torch/_refs/__init__.py:2963]: do you need a separate decomposition for this one? y[idx] translates into slice + index_put call without extra copies:
```
In [8]: def fn(x, y, z):
   ...:     idx = (slice(None),) + (z,)
   ...:     x[idx]=y
   ...: 

In [9]: z=torch.arange(3)

In [10]: with LoggingMode():
    ...:     fn(x,y,z)
    ...: 
$1: f32[4, 4] = aten.slice.Tensor($0, 0, 0, 9223372036854775807)
$1: f32[4, 4] = aten.index_put_.default($1, [None, $2], $3)
```

Comment (Lezcano) [reply torch/_refs/__init__.py:2963]: This one we can simply dispatch to `index_put_` directly so we're saving one kernel, I guess. Not that it changes too much. My reasoning was to have separate decompositions for all the setters.

Comment (ngimel) [reply torch/_refs/__init__.py:2963]: `slice` is op on strides, so we don't save a kernel. It's better to not have separate decompositions if we can avoid it.

File: test/test_ops.py [hunk 5]
```diff
@@ -235,7 +235,9 @@ def _ref_test_helper(self, ctx, device, dtype, op, skip_zero_numel=False, skip_z
                 if isinstance(a, torch.Tensor) or isinstance(b, torch.Tensor):
                     prims.utils.compare_tensor_meta(a, b)
                     if getattr(op, 'validate_view_consistency', True):
-                        self.assertEqual(a._is_view(), b._is_view())
+                        msg = (f"The torch implementation {'returns' if b._is_view() else 'does not return'} "
```
Comment (mruberry) [inline test/test_ops.py:238]: Nice message improvement

File: torch/_refs/__init__.py (hunk 4 above)
Comment (Lezcano) [reply torch/_refs/__init__.py:2963]: That's fair. I'll roll that one back. I'm just leaving `index_add` as an independent decomposition.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -17443,6 +17449,45 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
...
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_errors'),)
+    ),
+    PythonRefInfo(
+        "_refs.index_copy",
+        torch_opinfo_name="index_copy",
+        # empty_strided
+        supports_nvfuser=False,
+        skips=(
+            # no _refs support for Tensor.__setitem__
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref'),)
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:17471]: ouch, so not a lot of testing for the new refs

**Review by ngimel — APPROVED** [review]

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:17471]: They are tested against aten via the aten executor, right? I had to fix very many edge cases for tensors of dimension 0 lol

**Comment by Lezcano** [comment]
@malfet @seemethere @huydhn given the active development of PrimTorch and so, I think it'd be good to move the test that failed in this PR (that is `test/test_jit_cuda_fuser.py`) to always run in every PR.

File: torch/_decomp/decompositions.py [hunk 7]
```diff
@@ -1559,6 +1559,32 @@ def maybe_mask(vals, length, range_max, adaptive, dim):
...
+        index.ndim <= 1,
+        lambda: f"Index should have dimension 1 or 0 (got {index.ndim})",
+    )
+    if alpha != 1:
+        python_type = utils.dtype_to_type(x.dtype)
+        utils.check(
+            utils.is_weakly_lesser_type(type(alpha), python_type),
+            lambda: f"alpha argument of type {type(alpha)} cannot be safely cast to type {python_type}!",
+        )
+        tensor = torch._prims.mul(tensor, alpha)
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:1582]: it's better to use `tensor*alpha` here, inductor doesn't have lowering for prims.mul (in general, use `torch.* ` ops in decomps, not prims.

**Comment by ngimel** [comment]
> @malfet @seemethere @huydhn given the active development of PrimTorch and so, I think it'd be good to move the test that failed in this PR (that is `test/test_jit_cuda_fuser.py`) to always run in every PR.

The test that failed failed because of TorchScript issues (normally they should be found with jit_consistensy test, I'm not sure why it's not run), test_nvfuser_correctness and test_nnc_correctness are very slow tests and are not affected by primTorch development, so it doesn't make sense to move them to PR CI.

File: torch/_decomp/decompositions.py (hunk 7 above)
Comment (Lezcano) [reply torch/_decomp/decompositions.py:1582]: I'll change this in my next PR

Comment (Lezcano) [reply torch/_decomp/decompositions.py:1582]: FWIW, here I was replicating the behaviour in `add`.

**Pull request closed without merging**